extendedcopy.go

```diff
@@ -0,0 +1,123 @@
+/*
+Copyright The ORAS Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package oras
+
+import (
+	"context"
+	"errors"
+
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"oras.land/oras-go/v2/content"
+	"oras.land/oras-go/v2/errdef"
+	"oras.land/oras-go/v2/internal/descriptor"
+)
+
+// ExtendedCopy copies the directed acyclic graph (DAG) that are reachable from
+// the given tagged node from the source TraceableTarget to the destination Target.
+// The destination reference will be the same as the source reference if the
+// destination reference is left blank.
+// Returns the descriptor of the tagged node on successful copy.
+func ExtendedCopy(ctx context.Context, src TraceableTarget, srcRef string, dst Target, dstRef string) (ocispec.Descriptor, error) {
```

Comment:
Just curious. What if we pass a generic `Target` as `src`? It will do normal `Copy()` if target is not a `UpEdgeFinder`, and do `ExtendedCopy()` if it is, isn't it?

Reply:
If so, we don't even need to worry about name like `TraceableTarget` and `TraceableStorage`.

Reply:
If we pass a `Target` which is not a `UpEdgeFinder` to `ExtendedCopy`, should we throw an error saying extended copy is not supported for this `Target`?

Reply:
In that case, we can fall back to regular `Copy()`.

Reply:
I think falling back to regular `Copy()` is confusing, as users expect it to do extended copy. If it can't do, it should let users know.

Reply:
I am thinking about a scenario that a user is copying an artifact from one target to another. That user wants the artifact together with the associated signatures. The user has to do reflection on the Target first to see if it supports ExtendedCopy or not. Precisely,

```go
var desc ocispec.Descriptor
var err error
if target, ok := target.(TraceableTarget); ok {
    desc, err = oras.ExtendedCopy(...)
} else {
    desc, err = oras.Copy(...)
}
if err != nil {
    // handle error
}
```

From UX perspective, it would be better if `ExtendCopy()` is interpreted as _copying the artifact as well as its associated items if present_. A regular `Target`, which does not support `UpEdgeFinder`, can be viewed as a `TraceableTarget` that always return no up edges on any nodes.

Reply:
If the user wants to copy the artifact together with the associated signatures, they should use `ExtendedCopy()`.
If they does not care about the signature, they just uses `Copy()`.
The user can decide which one to use by themself.